mlir-clang/Lib/clang-mlir.cc

```diff
@@ -5293,69 +5297,65 @@ static bool parseMLIR(const char *Argv0, std::vector<std::string> filenames,
   const char *binary = Argv0; // CudaLower ? "clang++" : "clang";
   const unique_ptr<Driver> driver(
       new Driver(binary, llvm::sys::getDefaultTargetTriple(), Diags));
-  std::vector<const char *> Argv;
-  Argv.push_back(binary);
-  for (auto a : filenames) {
-    char *chars = (char *)malloc(a.length() + 1);
-    memcpy(chars, a.data(), a.length());
-    chars[a.length()] = 0;
-    Argv.push_back(chars);
-  }
+  std::vector<const char *> Argv = {binary};
+
+  for (const auto &a : filenames)
+    Argv.push_back(a.c_str());
+
   if (CudaLower)
     Argv.push_back("--cuda-gpu-arch=sm_35");
+
   if (FOpenMP)
     Argv.push_back("-fopenmp");
+
   if (Standard != "") {
     auto a = "-std=" + Standard;
     char *chars = (char *)malloc(a.length() + 1);
     memcpy(chars, a.data(), a.length());
     chars[a.length()] = 0;
     Argv.push_back(chars);
   }
+
   if (ResourceDir != "") {
     Argv.push_back("-resource-dir");
     char *chars = (char *)malloc(ResourceDir.length() + 1);
     memcpy(chars, ResourceDir.data(), ResourceDir.length());
     chars[ResourceDir.length()] = 0;
     Argv.push_back(chars);
   }
-  if (Verbose) {
+
+  if (Verbose)
     Argv.push_back("-v");
-  }
+
   if (CUDAGPUArch != "") {
     auto a = "--cuda-gpu-arch=" + CUDAGPUArch;
     char *chars = (char *)malloc(a.length() + 1);
     memcpy(chars, a.data(), a.length());
     chars[a.length()] = 0;
     Argv.push_back(chars);
   }
+
   if (MArch != "") {
     auto a = "-march=" + MArch;
     char *chars = (char *)malloc(a.length() + 1);
     memcpy(chars, a.data(), a.length());
     chars[a.length()] = 0;
     Argv.push_back(chars);
   }
-  for (auto a : includeDirs) {
+
+  for (const auto &a : includeDirs) {
     Argv.push_back("-I");
-    char *chars = (char *)malloc(a.length() + 1);
-    memcpy(chars, a.data(), a.length());
-    chars[a.length()] = 0;
-    Argv.push_back(chars);
+    Argv.push_back(a.c_str());
   }
-  for (auto a : defines) {
-    char *chars = (char *)malloc(a.length() + 3);
-    chars[0] = '-';
-    chars[1] = 'D';
-    memcpy(chars + 2, a.data(), a.length());
-    chars[2 + a.length()] = 0;
-    Argv.push_back(chars);
+
+  for (const auto &a : defines) {
+    Argv.push_back("-D");
+    Argv.push_back(a.c_str());
```

Comment:
The c_str here is illegal since the c_str doesn't necessarily outlast the for loop in scope [thus leading to later invalid reads].

Reply:
Why do you want the c_str() to outlast the for loop? You can still read "defines" later on if needed. Here I was following what Clang does see: clang/unittests/Driver/SanitizerArgsTest.cpp

Reply:
Perhaps I'm being over conservative, but I have run into issues where within clang not doing this results in reading undefined memory. From the [spec](https://www.cplusplus.com/reference/string/string/c_str/), this will no longer be valid after any modifiers to the string are run. If the for loop was pass by value, and not by reference, the deconstructor would be run invalidating it. However, since we're pass by value I suppose we're okay so long as no other part of the code changes the string or defines. As the defines vector is declared by value at the top of this function this means that any use of this string will be invalid after this function returns. Thus if this string is used internally as say a filename for the module, it would break. I can see why it doesn't break right now but it's sufficiently fragile that I'd prefer doing the memory allocation which is always safe. Open to thoughts/suggestions, though.

Reply:
I'm not worried about an actual string modification, but about the string deconstructor being called at one point (which I believe would invalidate it as well). So the worry is that the buildCompliation call could potentially capture the string, and use it after a deconstructor is called (thus leading to an undefined memory bug). I do think we're okay here right now though.